torchrl/envs/transforms/rlhf.py

```diff
@@ -0,0 +1,185 @@
+        >>> from torchrl.modules.distributions import TanhNormal
+        >>> from torch import nn
+        >>> base_env = GymEnv("Pendulum-v1")
+        >>> n_obs = base_env.observation_spec["observation"].shape[-1]
+        >>> n_act = base_env.action_spec.shape[-1]
+        >>> module = Mod(
+        ...     nn.Sequential(nn.Linear(n_obs, n_act * 2), NormalParamExtractor()),
+        ...     in_keys=["observation"],
+        ...     out_keys=["loc", "scale"],
+        ... )
+        >>> actor = ProbabilisticActor(
+        ...     module,
+        ...     in_keys=["loc", "scale"],
+        ...     distribution_class=TanhNormal,
+        ...     return_log_prob=True,
+        ... )
+        >>> transform = KLRewardTransform(actor, out_keys="reward_kl")
+        >>> env = TransformedEnv(base_env, transform)
+        >>> with torch.no_grad():
+        ...     # modify the actor parameters
+        ...     _ = TensorDict(dict(actor.named_parameters()), []).apply_(lambda x: x.data.copy_(x.data + 1))
+        ...     td = env.rollout(3, actor)
+        >>> # check that rewards have been modified
+        >>> assert (td.get(("next", "reward")) != td.get(("next", "reward_kl"))).all()
+
+    .. note::
+      Because the KL formulat is not always available and the parameters of the
+      original distribution may not have been recorded, we use a stochastic estimate
+      of the KL divergence.
+
+    """
+
+    DEFAULT_IN_KEYS = ["reward"]
+
+    def __init__(
+        self,
+        actor: ProbabilisticTensorDictModule,
+        coef=1.0,
+        in_keys=None,
+        out_keys=None,
+    ):
+        if in_keys is None:
+            in_keys = self.DEFAULT_IN_KEYS
+        if out_keys is None:
+            out_keys = in_keys
+        if not isinstance(in_keys, list):
+            in_keys = [in_keys]
+        if not isinstance(out_keys, list):
+            out_keys = [out_keys]
+        if not is_seq_of_nested_key(in_keys) or not is_seq_of_nested_key(out_keys):
+            raise ValueError(
+                f"invalid in_keys / out_keys:\nin_keys={in_keys} \nout_keys={out_keys}"
+            )
+        if len(in_keys) != 1 or len(out_keys) != 1:
+            raise ValueError(
+                f"Only one in_key/out_key is allowed, got in_keys={in_keys}, out_keys={out_keys}."
+            )
+        super().__init__(in_keys=in_keys, out_keys=out_keys)
+        # for convenience, convert out_keys to tuples
+        self.out_keys = [
+            out_key if isinstance(out_key, tuple) else (out_key,)
+            for out_key in self.out_keys
+        ]
+
+        # update the in_keys for dispatch etc
+        self.in_keys = self.in_keys + actor.in_keys
+
+        # check that the model has parameters
+        params = make_functional(
+            actor, keep_params=False, funs_to_decorate=["forward", "get_dist"]
+        )
+        self.functional_actor = deepcopy(actor)
+        repopulate_module(actor, params)
```

Comment:
Why do we repopulate here when below we seem to only call with `params=self.frozen_params`? Is this so that the caller can continue to use `actor` having supplied it as an argument to the constructor?